hypervisor/src/mshv/mod.rs

```diff
@@ -52,6 +51,205 @@ pub struct HvState {
 
 pub use HvState as VmState;
 
+struct IrqfdCtrlEpollHandler {
+    vm: Arc<dyn vm::Vm>, /* For issuing hypercall */
+    irqfd: EventFd,      /* Registered by caller */
+    kill: EventFd,       /* Created by us, signal thread exit */
+    epoll_fd: RawFd,     /* epoll fd */
+    gsi: u32,
+    gsi_routes: Arc<RwLock<HashMap<u32, MshvIrqRoutingEntry>>>,
+}
+
+fn register_listener(
+    epoll_fd: RawFd,
+    fd: RawFd,
+    ev_type: epoll::Events,
+    data: u64,
+) -> std::result::Result<(), io::Error> {
+    epoll::ctl(
+        epoll_fd,
+        epoll::ControlOptions::EPOLL_CTL_ADD,
+        fd,
+        epoll::Event::new(ev_type, data),
+    )
+}
+
+const KILL_EVENT: u16 = 1;
+const IRQFD_EVENT: u16 = 2;
+
+impl IrqfdCtrlEpollHandler {
+    fn run_ctrl(&mut self) {
+        self.epoll_fd = epoll::create(true).unwrap();
+        let epoll_file = unsafe { File::from_raw_fd(self.epoll_fd) };
+
+        register_listener(
+            epoll_file.as_raw_fd(),
+            self.kill.as_raw_fd(),
+            epoll::Events::EPOLLIN,
+            u64::from(KILL_EVENT),
+        )
+        .unwrap_or_else(|err| {
+            info!(
+                "IrqfdCtrlEpollHandler: failed to register listener: {:?}",
+                err
+            );
+        });
+
+        register_listener(
+            epoll_file.as_raw_fd(),
+            self.irqfd.as_raw_fd(),
+            epoll::Events::EPOLLIN,
+            u64::from(IRQFD_EVENT),
+        )
+        .unwrap_or_else(|err| {
+            info!(
+                "IrqfdCtrlEpollHandler: failed to register listener: {:?}",
+                err
+            );
+        });
```

Comment:
I would let @liuw  to comment.

Reply:
No. I don't think it makes `run_ctrl()` exit. The thread will be just hanging forever because it will block on those eventfds.

I was not sure at the time there would be a better way to handle this. I didn't want to exit the thread because I was worried that it might cause failure when `unregister_irqfd` wrote to the `kill_fd`. Now that I think about this, it is probably okay to let the thread exit because `kill_fd` will hang on for as long as needed.

Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

Reply:
> Note that the emulation is just a stop-gap. It is going to be superseded by our work to implement irqfd and ioeventfd inside Linux kernel. We already wrote some patches and they will be posted once ready.

Then let's keep it as is for now.